Seed RNG with precision timestamps on receipt of net messages.

TheBlueMatt opened the pull request:
See title. Exposes a generic dead-simple "SeedEvent" interface, but currently just used for net messages.

TheBlueMatt committed:
Seed RNG with precision timestamps on receipt of net messages.

sipa commented on src/random.h:
```diff
@@ -89,6 +89,14 @@ void GetStrongRandBytes(unsigned char* buf, int num) noexcept;
+ *
+ * Thread-safe.
+ */
+void SeedEvent(const unsigned char* event_type_buf, size_t buf_len);
```
Do we expect the argument to always be a string? The interface could be made a bit simpler and safer if that's the case (you could also avoid the size_t prefix, and instead just write the c_str() including null byte, which self-describes the length).

TheBlueMatt replied:
I presume not, both for upcoming p2p encryption where the message type may just be a char, and I currently have some radio stuff hooked up that dumps RSSI information into this.

laanwj replied:
I kind of prefer this general (ptr,size) interface to using C strings, whose use we're generally trying to avoid in bitcoin code. `std::slice` would be a safer interface, if only C++11 had it.

sipa replied:
Here is a more C++ish interface, that serializes (through the serialization framework) all arguments passed to SeedEvent (without any allocations): https://github.com/sipa/bitcoin/tree/bluematt_2019-11-rng-netmsg

TheBlueMatt replied:
Oof, that's....quite a bigger diff. Are you worried about the current interface introducing bugs? I don't think it's particularly likely (nor are we likely to add calls to this *everywhere*), and it matches the rest of the random. interface.

sipa replied:
If you switch to SHA256 (which you did) it's much smaller.

It also enables efficiently adding extra information without first copying everything into a buffer (say, you could call `SeedEvent(msg.command, peerid);` just as easily)

sipa commented on src/random.cpp:
```diff
@@ -485,6 +488,14 @@ static void SeedPeriodic(CSHA512& hasher, RNGState& rng)
     // High-precision timestamp
     SeedTimestamp(hasher);
 
+    {
```
If you'd abstract this code block out, and then call it from say both `SeedPeriodic` and `SeedSlow`, the additional entropy would always be available to strong RNG output, even if a minute has not yet passed. You could make the hashing step conditional on `events_hasher.Size() != 0`.

TheBlueMatt replied:
Hmm, you mean in SeedSlow Finalize the events_hasher, then write the buffer back into the events_hasher and have SeedPeriodic wipe it? Cause SeedSlow doesn't store the data back into the RNGState.

sipa replied:
All the Seed* functions just add to a CSHA512; ultimately, they're all called from ProcRand, which mixes it with the existing RNGState, and produces output.

SeedPeriodic goes a step further, in trying to make sure that the existing RNGState is also fed through the strengthener, but that isn't really required.

TheBlueMatt replied:
Right, so I did this.

TheBlueMatt commented:
It was pointed out that this would make processing of small net messages particularly slower. Instead of only mixing in extra randomness for large net messages I switched it to SHA256 to use our existing ASM optimizations here. Also changed the fn name to be closer to the rest of random.h.

sipa commented on src/random.cpp:
```diff
@@ -553,6 +577,16 @@ void GetRandBytes(unsigned char* buf, int num) noexcept { ProcRand(buf, num, RNG
 
+void RandAddEvent(const unsigned char* event_type_buf, size_t buf_len) {
+    LOCK(events_mutex);
+    events_hasher.Write((unsigned char *)&buf_len, sizeof(size_t));
```
You could also limit this to 4 bytes (or even 2...), if size is an issue.

TheBlueMatt replied:
Hmm, I'm not too worried about it? Its mostly up to the caller and right now net will only ever pass in 12+4+8 bytes per call, which means it will never be more than one extra hash per two messages, which seems reasonable.

sipa commented on src/random.cpp:
```diff
@@ -553,6 +577,16 @@ void GetRandBytes(unsigned char* buf, int num) noexcept { ProcRand(buf, num, RNG
+    events_hasher.Write(event_type_buf, buf_len);
+    // Get the low four bytes of the performance counter. This translates to roughly the
+    // subsecond part.
+    int32_t perfcounter = (GetPerformanceCounter() & 0xffffffff);
```
Nit: `uint32_t` is probably more appropriate

sipa commented:
A comment on including the timestamp: GetPerformanceCounter on ARM is not extremely cheap (it involves a system call), perhaps you'd want a way for the caller to provide their own timestamp (or timestamp-like variable), as I assume generally they already have one?

TheBlueMatt commented:
Hmm, net does have one, but its also mostly intended for things exactly like net, where you are handling an "event" that is the result of a syscall already...Unless its particularly more expensive than poll() + read(), I'm not too worried :). Its not like we're gonna call it somewhere where performance in nanoseconds is absolutely critical, though may be calling it in places where DoS is a risk.

sipa commented:
I would indeed assume that it's not more expensive than poll+read, good point.

sipa commented:
Here is another idea: use the serialization-based interface, but instead feeding it directly into events_hasher, feed it into a SipHasher instead. Then feed a few bits (8, 16, 32?) of the SipHash of all passed data to events_hasher. That should be both simpler (no need to expose internals as much), and faster (much less SHA256'ing, don't hold lock as long).

TheBlueMatt commented:
Even simpler idea, just use the damn message hash instead of the command :). Its nice and small (4 bytes), plus has more entropy in it anyway. Since our (real) goal is to get the time when we receive a message, that's more than sufficient, but we also steal our peers' entropy in their ping nonces :).

sipa commented:
Nice... if the caller has too much data they can always siphash or wahtever it thenselves. You could simplify the interface by just allowing a uint32_t to be passed or so.

TheBlueMatt commented:
Done. Its now an explicit uint32_t.

sipa commented on src/net.cpp:
```diff
@@ -693,6 +697,10 @@ CNetMessage V1TransportDeserializer::GetMessage(const CMessageHeader::MessageSta
     msg.m_raw_message_size = hdr.nMessageSize + CMessageHeader::HEADER_SIZE;
 
+    // We just received a message off the wire, harvest entropy from the time (and the message checksum)
+    uint32_t msg_hash = (hash.begin()[0] << 8*3) | (hash.begin()[1] << 8*2) | (hash.begin()[2] << 8*1) | hash.begin()[3];
```
`ReadLE32(hash.begin())` ?

sipa commented on src/random.h:
```diff
@@ -89,6 +89,14 @@ void GetStrongRandBytes(unsigned char* buf, int num) noexcept;
 
+/**
+ * Gathers entropy from the low bits of the time at which events occur. Should
+ * be called with a few bytes describing the event at the time an event occurs.
```
This comment seems outdated now.

TheBlueMatt replied:
Heh, I figured I'd leave it, since it doesn't actually matter what the 4 bytes are, they just have to be aligned and typed as a u32.

sipa commented on src/random.cpp:
```diff
@@ -449,6 +450,23 @@ static void SeedFast(CSHA512& hasher) noexcept
+    events_hasher.Finalize(events_hash);
+    hasher.Write(events_hash, 32);
+
+    // Re-initialize the hasher with the finalized state to use later.
```
Ha, neat.

sipa commented:
utACK https://github.com/bitcoin/bitcoin/commit/02d8c56a18b9a2960888d6ec1209955105bae847

sipa commented:
This effectively adds one SHA256 compression per 8 messages/connections. That seems acceptably low, compared to the work already done for message checksumming.

meshcollider reviewed (commented):
utACK 02d8c56a18b9a2960888d6ec1209955105bae847

TheBlueMatt commented:
The AppVeyor build fail occurs before build even starts, so is a false positive.

practicalswift commented:
Concept ACK: nice idea and straightforward implementation

practicalswift commented:
Clang Thread Safety Analysis annotations seem to be missing? :)

naumenkogs commented:
Concept ACK

laanwj commented:
ACK 02d8c56a18b9a2960888d6ec1209955105bae847

practicalswift commented:
@laanwj I was a bit surprised to see this merged without thread safety annotations, or more specifically before the thread safety questions were addressed. Was there a reason for that? :)

laanwj commented:
It has three ACKs and passes travis. 

If you want to add any annotations, go ahead, but we're not going to block merges on always new requirements.

sipsorcery commented:
This change is causing problems on Windows. No appveyor tests have passed since it was merged and I've also verified locally that  `bitcoind.exe`, `bitcoin-qt.exe`, `bench_bitcoin.exe` and `test_bitcoin.exe` are failing. Failing in this case means exiting almost immediately with no error message.

Running the programs in the Visual Studio debugger gets an `Access Violation` exception reported on the `LOCK(events_mutex)` call. I'm always cautious with `Access Violation` exceptions as often the cause could have been an earlier call and the reported site is a side effect.

As a quick and dirty test I moved the declaration of `static Mutex events_mutex` to be a local static variable in the two methods using it. The programs were all able to run normally after that. I realise that's defeating the purpose of the mutex so isn't a fix.

sipa commented:
@sipsorcery I should have known that would happen; I suspect it is invoking the RNG before global initialization is complete, and thus potentially before events_mutex is initialized. Working on a fix.
